src/ucp/core/ucp_rkey.c

```diff
@@ -213,61 +218,507 @@ UCS_PROFILE_FUNC(ssize_t, ucp_rkey_pack_memh,
 }
 
+static size_t ucp_memh_extended_info_size(size_t common_size,
+                                          size_t specific_info_size)
+{
+    if ((specific_info_size + common_size - sizeof(uint8_t)) > UINT8_MAX) {
+        /* This field is packed after the 8-bit size field in case of a total
+         * info size is greater than UINT8_MAX. */
+        return sizeof(uint16_t);
+    }
+
+    return 0;
+}
+
+static size_t ucp_memh_common_packed_size(size_t specific_info_size)
+{
+    size_t size;
+
+    /* Size of mkey information which comes prior TL mkey data of all MDs */
+    size = sizeof(uint8_t);
+
+    /* Flags */
+    size += sizeof(uint16_t);
+
+    /* Memory domains map */
+    size += sizeof(uint64_t);
+
+    /* Memory type */
+    size += sizeof(uint8_t);
+
+    return size + ucp_memh_extended_info_size(size, specific_info_size);
+}
+
+static size_t ucp_memh_exported_info_packed_size()
+{
+    size_t size;
+
+    /* Address */
+    size = sizeof(uint64_t);
+
+    /* Length */
+    size += sizeof(uint64_t);
+
+    /* UCP uuid */
+    size += sizeof(uint64_t);
+
+    /* Registration ID */
+    size += sizeof(uint64_t);
+
+    /* Common data (size, md_map, mem_type and etc) */
+    size += ucp_memh_common_packed_size(size);
+
+    return size;
+}
+
+static void ucp_memh_info_size_pack(void **p, size_t info_size)
+{
+    ucs_assert(info_size <= UINT16_MAX);
+
+    if ((info_size - sizeof(uint8_t)) > UINT8_MAX) {
+        *ucs_serialize_next(p, uint8_t)  = 0;
+        *ucs_serialize_next(p, uint16_t) = info_size - sizeof(uint16_t);
+    } else {
+        *ucs_serialize_next(p, uint8_t) = info_size - sizeof(uint8_t);
+    }
+}
+
+uint16_t ucp_memh_info_size_unpack(const void **p)
+{
+    uint16_t info_size = *ucs_serialize_next(&p, uint8_t);
+
+    if (info_size == 0) {
+        info_size = *ucs_serialize_next(&p, uint16_t) + sizeof(uint16_t);
+    }
+
+    return info_size + sizeof(uint8_t);
+}
+
+static void ucp_memh_common_pack(const ucp_mem_h memh, void **p,
+                                 uint64_t flags, size_t memh_info_size)
+{
+    ucp_context_h UCS_V_UNUSED context = memh->context;
+
+    /* Check that md_map is valid */
+    ucs_assertv(ucs_test_all_flags(UCS_MASK(context->num_mds), memh->md_map),
+                "mask 0x%lx memh %p md_map 0x%lx", UCS_MASK(context->num_mds),
+                memh, memh->md_map);
+
+    /* Size of mkey information which comes prior TL mkey data of all MDs */
+    ucp_memh_info_size_pack(p, memh_info_size);
+
+    /* Flags */
+    *ucs_serialize_next(p, uint16_t) = flags;
+
+    /* Memory domains map */
+    *ucs_serialize_next(p, uint64_t) = context->export_md_map[memh->mem_type];
+
+    /* Memory type */
+    UCS_STATIC_ASSERT(UCS_MEMORY_TYPE_LAST <= 255);
+    *ucs_serialize_next(p, uint8_t) = memh->mem_type;
+}
+
+size_t ucp_memh_global_id_packed_size(uct_md_attr_v2_t *md_attr)
+{
+    size_t size = UCT_MD_GLOBAL_ID_MAX;
+
+    while ((size != 0) && (md_attr->global_id[size - 1] == '\0')) {
```

Comment:
why not going from the beginning. Is there any guarantee that unused part of global_id is set to specific symbols (e.g. `\0`)

Reply:
there are no guarantees that no `\0` at the beginning or middle of the globail_id buffer filled by UCT transports.
but if we have K `\0` symbols starting from the end, we can just not pack them